driver_test.go

```diff
@@ -12,26 +12,263 @@
+		return err
+	}
+	if err := op.Wait(curs.ctx); err != nil {
+		return err
 	}
-	connector, err := driver.OpenConnector("projects/$PROJECT/instances/$INSTANCE/databases/$DATABASE")
+	return nil
+}
+
+// Executes DML using the client library.
+func ExecuteDMLClientLib(dml []string) (err error) {
+
+	// Open client/
+	ctx := context.Background()
+	client, err := spanner.NewClient(ctx, dsn)
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	// Put strings into spanner.Statement structure.
+	var stmts []spanner.Statement
+	for _, line := range dml {
+		stmts = append(stmts, spanner.NewStatement(line))
+	}
+
+	// Execute statements.
+	_, err = client.ReadWriteTransaction(ctx, func(ctx context.Context, txn *spanner.ReadWriteTransaction) error {
+		stmts := stmts
+		rowCounts, err := txn.BatchUpdate(ctx, stmts)
+		if err != nil {
+			return err
+		}
+		fmt.Printf("Executed %d SQL statements using Batch DML.\n", len(rowCounts))
+		return nil
+	})
+
+	return err
+}
+
+func TestQueryContext(t *testing.T) {
+
+	// Set up test table.
+	curs, err := NewConnector()
 	if err != nil {
 		log.Fatal(err)
 	}
-	db := sql.OpenDB(connector)
-	_ = db // Use db.
+	defer curs.Close()
+
+	err = executeDdlApi(curs, []string{`CREATE TABLE TestQueryContext (
+		A   STRING(1024),
+		B  STRING(1024),
+		C   STRING(1024)
+	)	 PRIMARY KEY (A)`})
+	if err != nil {
+		t.Error(err)
+	}
+	err = ExecuteDMLClientLib([]string{`INSERT INTO TestQueryContext (A, B, C) 
+		VALUES ("a1", "b1", "c1"), ("a2", "b2", "c2") , ("a3", "b3", "c3") `})
+	if err != nil {
+		t.Error(err)
+	}
+
+	type testQueryContextRow struct {
+		A, B, C string
+	}
+
+	tests := []struct {
+		input          string
+		want           []testQueryContextRow
+		wantErrorQuery string
+		wantErrorScan  string
+	}{
+		// empty query
+		{input: "", want: []testQueryContextRow{}},
+		// syntax error
+		{input: "SELECT SELECT * FROM TestQueryContext", want: []testQueryContextRow{}},
+		// retur nothing
+		{input: "SELECT SELECT * FROM TestQueryContext", want: []testQueryContextRow{}},
+		// return one tuple
+		{input: "SELECT * FROM TestQueryContext WHERE A = \"a1\"",
+			want: []testQueryContextRow{
+				{A: "a1", B: "b1", C: "c1"},
+			}},
+		// return subset of tuples
+		{input: "SELECT * FROM TestQueryContext WHERE A = \"a1\" OR A = \"a2\"",
+			want: []testQueryContextRow{
+				{A: "a1", B: "b1", C: "c1"},
+				{A: "a2", B: "b2", C: "c2"},
+			}},
+		// subet of tuples with !=
+		{input: "SELECT * FROM TestQueryContext WHERE A != \"a3\"",
+			want: []testQueryContextRow{
+				{A: "a1", B: "b1", C: "c1"},
+				{A: "a2", B: "b2", C: "c2"},
+			}},
+		// return entire table
+		{input: "SELECT * FROM TestQueryContext ORDER BY A",
+			want: []testQueryContextRow{
+				{A: "a1", B: "b1", C: "c1"},
+				{A: "a2", B: "b2", C: "c2"},
+				{A: "a3", B: "b3", C: "c3"},
+			}},
+		// query non existant table
+		{input: "SELECT * FROM TestQueryContexta", want: []testQueryContextRow{}},
+	}
+
+	// Open db.
```

Comment:
nit: I would move this section to be below the DML as it is part of that setup and there's no reason to separate declaring the test cases from iterating over them.

Reply:
moved 📦